### PR TITLE
[player,wscript]: os:win cplugins support

### PR DESCRIPTION
**Pull request by 57op**
Hi,
I was wondering why there wasn't cplugins support for windows,
so I gave it a try and it worked fine with a toy extension I'm developing.

I'm not very familiar with WAF build system, I hope the changes are appreciated.
About `scripting.c`, the `HAVE_CPLUGINS` section is not the best looking code (many #if/#else/#endif)... if you know or want any pattern to be applied I'll gladly try to refactor.

I only tested building and running on windows target with this changes, other platforms should need a test.

### CODE REVIEW

**Comment by wm4** [comment]
CC @rossy

**Comment by rossy** [comment]
So, the problem with the current cplugin API on Windows is that the Windows linker doesn't have global symbols. It always links to a .dll (or .exe) name, which means a plugin that links to the client API in mpv.exe would have to be recompiled to be used in a libmpv app, because it would have to link to mpv-1.dll instead. Similarly, if someone renames their mpv.exe for whatever reason, cplugins would stop working for them, and if someone statically links libmpv to their frontend, the cplugins would have to link to their frontend's .exe name.

This is more-or-less the reason why we haven't enabled cplugins on Windows yet. I'm not sure how to solve that problem. I think there might be a PE feature that lets an image export functions under a different name to its file name, which might be how drivers link to kernel APIs in both ntoskrnl.exe and ntkrnlmp.exe. I don't know if that feature would work for us though.

Another way to do it would be to have a plugin SDK that exposes stubs for all the libmpv functions and finds them at runtime. We could also have a v2 of the cplugin API that passes a struct to mpv_open_cplugin containing function pointers to the client API.

We could just merge this as-is, with the caveat that plugins will only work with mpv.exe (with the original file name) and not with mpv-1.dll. That might be better than nothing, though I'm worried it will create an ecosystem of Windows cplugins that will all have to be changed if we change the cplugin API to something that solves the linking problem.

**Comment by wm4** [comment]
> We could also have a v2 of the cplugin API that passes a struct to mpv_open_cplugin containing function pointers to the client API.

That wouldn't work, because then there'd be 2 APIs around.

Another possibility is making mpv.exe a stub that just links against libmpv.dll. From what I understand, there'd be no more symbol problems.

**Comment by chuck-** [comment]
> 
> 
> > We could also have a v2 of the cplugin API that passes a struct to mpv_open_cplugin containing function pointers to the client API.
> 
> That wouldn't work, because then there'd be 2 APIs around.
> 
> Another possibility is making mpv.exe a stub that just links against libmpv.dll. From what I understand, there'd be no more symbol problems.

Indeed, I've just done a quick attempt and it works as expected.
The only difference, of course, is that mpv-1.dll against which mpv.exe is linked must export more symbols than just the defined ones in mpv.def (I used the ld flag `-Wl,--export-all-symbols` in my attempt).

I don't know if this solves the issue mentioned by @rossy whether a plugin should need to be recompiled if targeting mpv or libmpv.
I guess it should, but I have no way to test it.

**Comment by wm4** [comment]
> is that mpv-1.dll against which mpv.exe is linked must export more symbols than just the defined ones in mpv.def

What? Why?

**Comment by chuck-** [comment]
> What? Why?

Oh well, never mind.
You're right, not needed.

**Comment by wm4** [comment]
I guess we could merge it (possibly forcing authors to provide 2 builds), so proper review follows.

File: player/scripting.c [hunk 1]
```diff
@@ -292,32 +292,57 @@ bool mp_load_scripts(struct MPContext *mpctx)
...
+    #if _WIN32
+    DWORD err = GetLastError();
+    if (err)
+        MP_ERR(args, "C plugin error: '%lu'\n", err);
+    #else
     char *err = dlerror();
     if (err)
         MP_ERR(args, "C plugin error: '%s'\n", err);
+    #endif
     return -1;
```
Comment (wm4) [inline player/scripting.c:335]: So I think it'd be better to either 1. duplicate the whole loader in a windows version, or 2. provide abstraction wrappers for dlopen etc. for windows. I picture 2. as providing a POSIX-like API on win32, probably messed into osdep/io.c.

File: waftools/checks/custom.py [hunk 2]
```diff
@@ -151,3 +151,13 @@ def fn(ctx, dependency_identifier, **kw):
             else:
                 return False
     return fn
+
+def check_cplugins_linkflags(ctx, dependency_identifier):
+    checks = [
+        check_cc(linkflags=['-rdynamic']),
+        check_cc(linkflags=['-Wl,--out-implib,libmpv.dll.a']) ]
```
Comment (wm4) [inline waftools/checks/custom.py:158]: How exactly does this work? Does it require a libmpv.dll.a to be present?

File: wscript [hunk 3]
```diff
@@ -968,16 +968,19 @@ def configure(ctx):
...
-
+    
     if ctx.dependency_satisfied('cplugins'):
         # We need to export the libmpv symbols, since the mpv binary itself is
         # not linked against libmpv. The C plugin needs to be able to pick
         # up the libmpv symbols from the binary. We still restrict the set
         # of exported symbols via mpv.def.
-        ctx.env.LINKFLAGS += ['-rdynamic']
+        if ctx.dependency_satisfied('mingw'):
+            ctx.env.LINKFLAGS += ['-Wl,--out-implib,libmpv.dll.a']
```
Comment (wm4) [inline wscript:981]: Same question here. (And why is it duplicated? I know it was duplicated before.)

File: waftools/checks/custom.py (hunk 2 above)
Comment (chuck-) [reply waftools/checks/custom.py:158]: No, it's for generating an import lib from the exe to be used for linking cplugins.
If you are going to follow the
>mpv.exe a stub that just links against libmpv.dll
approach it won't be needed.

By the way, I noticed this breaks the build process when enabling `--enable-libmpv-shared`.
I'm sorry for the mess, I'm very unfamiliar with waf and should have thought better before pulling this request.

File: player/scripting.c (hunk 1 above)
Comment (chuck-) [reply player/scripting.c:335]: For 2. do you mean something an abstraction like mpv_{dlopen,dlsym,dlerror}?
Or defining dlopen, dlysym and dlerror on win32 only?

Anyway, I recently noticed that DLLs with non ascii characters fails to load.
I will switch from LoadLibraryA to LoadLibraryW

File: wscript (hunk 3 above)
Comment (chuck-) [reply wscript:981]: I replicated the behavior in that script (`-rdynamic` has been duplicated as well).

I intend to revert these changes anyway.
They're not needed if we build mpv.exe as a stub that links against libmpv.dll.

File: player/scripting.c (hunk 1 above)
Comment (wm4) [reply player/scripting.c:335]: Basically, the latter, but see io.h and io.c.

Comment (chuck-) [reply player/scripting.c]: Okay, I've uploaded those changes.
For the time being, I won't touch anything related to the build system even though cplugins autodetection could be enabled, provided we disable `-rdynamic` compiler option on windows.

Furthermore, I don't know if already present (dynamic build aside), we should introduce an option to build mpv.exe as a stub for mpv-1.dll.

File: osdep/io.c [hunk 4]
```diff
@@ -692,6 +692,58 @@ off_t mp_lseek(int fd, off_t offset, int whence)
...
+    
+    return (void *)LoadLibraryW(wfilename);
+}
+
+void *mp_dlsym(void *handle, const char *symbol)
+{
+  return (void *)GetProcAddress((HMODULE)handle, symbol);
+}
+
+static char mp_dlerror_buffer[250];
```
Comment (wm4) [inline osdep/io.c:720]: Right, it seems dlerror() can unfortunately not be implemented without global state. But it should use TLS, try adding _Thread_local to the declaration.

File: osdep/io.c [hunk 5]
```diff
@@ -692,6 +692,58 @@ off_t mp_lseek(int fd, off_t offset, int whence)
     return _lseeki64(fd, offset, whence);
 }
 
+void *mp_dlopen(const char *filename, int flag)
```
Comment (wm4) [inline osdep/io.c:695]: Not sure if any of the flags could/should be translated to winapi equivalents, CC @rossy.

File: osdep/io.c [hunk 6]
```diff
@@ -692,6 +692,58 @@ off_t mp_lseek(int fd, off_t offset, int whence)
...
+void *mp_dlsym(void *handle, const char *symbol)
+{
+  return (void *)GetProcAddress((HMODULE)handle, symbol);
+}
+
+static char mp_dlerror_buffer[250];
+
+char *mp_dlerror(void)
+{
+    DWORD err = GetLastError();
```
[the remark this answers is not in the document]
Comment (chuck-) [reply osdep/io.c:724]: Adding _Thread_local to `err` declaration yields a compiler error:
> error: function-scope 'err' implicitly auto and declared '_Thread_local'

File: osdep/io.c (hunk 4 above)
Comment (chuck-) [reply osdep/io.c:720]: Okay, it compiles correctly here

File: osdep/io.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (wm4) [reply osdep/io.c:724]: It has to be a static/global variable like mp_dlerror_buffer. The important bit is that the error value of LoadLibrary should be preserved.

[the remark this answers is not in the document]
Comment (chuck-) [reply osdep/io.c:724]: I got it.
So, should I save last error code soon after LoadLibrary and GetProcAddress are invoked?

```
_Thread_local static DWORD mp_dlerror_code;  
  
void *mp_dlopen(...)
     LoadLibrary(...)
     mp_dlerror_code = GetLastError()

void *mp_dlsym(...)
     GetProcAddress(...)
     mp_dlerror_code = GetLastError()
```

[the remark this answers is not in the document]
Comment (wm4) [reply osdep/io.c:724]: Yes.

[the remark this answers is not in the document]
Comment (chuck-) [reply osdep/io.c:724]: Fine, I'll wait on rossy feedback about dlopen flags and update once again.

File: osdep/io.c (hunk 5 above)
Comment (chuck-) [reply osdep/io.c]: I found this dlfcn implementation for win32: https://github.com/dlfcn-win32
They try to translate some of the RTLD_ flags, e.g. RTLD_LOCAL: https://github.com/dlfcn-win32/dlfcn-win32/blob/master/dlfcn.c#L286
by keeping a doubly linked list of opened libraries handles.
Given that we don't generally dlclose opened libraries, I guess the behavior should be the same.

As far as I understand, RTLD_NOW is the default behavior on windows.
There's a way to not make it resolve symbols at all (LoadLibraryEx + DONT_RESOLVE_DLL_REFERENCES flag) which is not the same as RTLD_LAZY (anyway, its use it's discouraged by the official documentation in favor of other flags for other purposes).

**Pull request closed without merging**